### PR TITLE
Refactor misc endpoints for user management

### PR DESCRIPTION
**Pull request by ivov**
/oauth1-credential/auth
/oauth2-credential/auth
/oauth1-credential/callback
/oauth2-credential/callback
/node-parameter-options
~~/user-survey~~
~~/test-webhook/:id~~
~~/data/:path~~

### CODE REVIEW

File: packages/cli/src/Server.ts [hunk 1]
```diff
@@ -1881,121 +1884,119 @@ class App {
...
+					const { id: credentialId } = req.query;
 
-				const queryBuilder = Db.collections.Credentials!.createQueryBuilder('c');
-				queryBuilder.andWhere('c.id = :id', { id: req.query.id });
-				queryBuilder.innerJoin('c.shared', 'shared');
-				queryBuilder.andWhere('shared.userId = :userId', {
-					userId: (req.user as User).id,
-				});
+					if (!credentialId) {
+						res.status(500).send('Required credential ID is missing!');
```
Comment (krynble) [inline packages/cli/src/Server.ts:1892]: We are using `ResponseHelper.send` in the function signature, so you can simply use `throw new ResponseHelper.ResponseError('Required credential ID is missing!', undefined, 500);` and it should work.

File: packages/cli/src/Server.ts [hunk 2]
```diff
@@ -1881,121 +1884,119 @@ class App {
...
+					const credential = await getCredentialForUser(credentialId, req.user);
 
-				let encryptionKey;
-				encryptionKey = await UserSettings.getEncryptionKey();
-				if (encryptionKey === undefined) {
-					res.status(500).send('No encryption key got found to decrypt the credentials!');
-					return '';
-				}
+					if (!credential) {
+						res.status(404).send('The credential is not known.');
```
Comment (krynble) [inline packages/cli/src/Server.ts:1899]: Same as above, throwing error already stops the function execution so it seem a bit more elegant than returning an empty string.

File: packages/cli/src/Server.ts [hunk 3]
```diff
@@ -1881,121 +1884,119 @@ class App {
...
-				const oauthCredentials = credentialsHelper.applyDefaultsAndOverwrites(
-					decryptedDataOriginal,
-					result.type,
-					mode,
-				);
+					const encryptionKey = await UserSettings.getEncryptionKey();
 
-				const signatureMethod = _.get(oauthCredentials, 'signatureMethod') as string;
+					if (!encryptionKey) {
+						res.status(500).send('No encryption key found to decrypt the credentials!');
```
Comment (krynble) [inline packages/cli/src/Server.ts:1906]: Again ^_^

File: packages/cli/src/Server.ts [hunk 4]
```diff
@@ -2103,104 +2098,102 @@ class App {
...
+					const { id: credentialId } = req.query;
 
-				const queryBuilder = Db.collections.Credentials!.createQueryBuilder('c');
-				queryBuilder.andWhere('c.id = :id', { id: req.query.id });
-				queryBuilder.innerJoin('c.shared', 'shared');
-				queryBuilder.andWhere('shared.userId = :userId', {
-					userId: (req.user as User).id,
-				});
+					if (!credentialId) {
+						res.status(500).send('Required credential ID is missing.');
```
Comment (krynble) [inline packages/cli/src/Server.ts:2106]: Also here

File: packages/cli/src/Server.ts [hunk 5]
```diff
@@ -2103,104 +2098,102 @@ class App {
...
+					const credential = await getCredentialForUser(credentialId, req.user);
 
-				let encryptionKey;
-				encryptionKey = await UserSettings.getEncryptionKey();
-				if (encryptionKey === undefined) {
-					res.status(500).send('No encryption key got found to decrypt the credentials!');
-					return '';
-				}
+					if (!credential) {
+						res.status(404).send('The credential is not known.');
```
Comment (krynble) [inline packages/cli/src/Server.ts:2113]: Yup

File: packages/cli/src/Server.ts [hunk 6]
```diff
@@ -2103,104 +2098,102 @@ class App {
...
-					accessTokenUri: _.get(oauthCredentials, 'accessTokenUrl', '') as string,
-					authorizationUri: _.get(oauthCredentials, 'authUrl', '') as string,
-					redirectUri: `${WebhookHelpers.getWebhookBaseUrl()}${
-						this.restEndpoint
-					}/oauth2-credential/callback`,
-					scopes: _.split(_.get(oauthCredentials, 'scope', 'openid,') as string, ','),
-					state: stateEncodedStr,
-				};
+					if (!encryptionKey) {
+						res.status(500).send('No encryption key found to decrypt the credentials!');
```
Comment (krynble) [inline packages/cli/src/Server.ts:2120]: Last one, I promise.

**Review by krynble — CHANGES_REQUESTED** [review]
Minor (and even optional) changes - let me know your thoughts.

File: packages/cli/src/Server.ts [hunk 7]
```diff
@@ -1967,121 +1974,124 @@ class App {
...
+					const credential = await getCredentialForUser(credentialId, req.user);
 
-				let encryptionKey;
-				encryptionKey = await UserSettings.getEncryptionKey();
-				if (encryptionKey === undefined) {
-					res.status(500).send('No encryption key got found to decrypt the credentials!');
-					return '';
-				}
+					if (!credential) {
+						throw new ResponseHelper.ResponseError('The credential is not known', undefined, 404);
```
Comment (BHesseldieck) [inline packages/cli/src/Server.ts:1992]:
```suggestion
						throw new ResponseHelper.ResponseError('Credential not found', undefined, 404);
```

File: packages/cli/src/Server.ts [hunk 8]
```diff
@@ -1246,35 +1253,35 @@ class App {
...
 						currentNodeParameters,
 						credentials,
 					);
 
 					const additionalData = await WorkflowExecuteAdditionalData.getBase(currentNodeParameters);
-					// TODO UM: restrict user access to credentials he cannot use.
-					// @ts-ignore
-					additionalData.userId = (req.user as User).id;
+
+					additionalData.userId = req.user.id;
```
Comment (BHesseldieck) [inline packages/cli/src/Server.ts:1284]: why we need to add this here?

File: packages/cli/src/Server.ts [hunk 9]
```diff
@@ -1246,35 +1253,35 @@ class App {
...
-						credentials = JSON.parse(req.query.credentials as string);
-					}
-					const methodName = req.query.methodName as string;
 
-					const nodeTypes = NodeTypes();
+					let credentials: INodeCredentials | undefined;
+
+					if (req.query.credentials) {
+						credentials = JSON.parse(req.query.credentials);
+						// TODO UM: restrict user access to credentials he cannot use.
```
Comment (BHesseldieck) [inline packages/cli/src/Server.ts:1271]: can be removed?
```suggestion
```

File: packages/cli/src/Server.ts [hunk 10]
```diff
@@ -1967,121 +1974,124 @@ class App {
...
-					decryptedDataOriginal,
-					result.type,
-					mode,
-				);
+					const encryptionKey = await UserSettings.getEncryptionKey();
 
-				const signatureMethod = _.get(oauthCredentials, 'signatureMethod') as string;
+					if (!encryptionKey) {
+						throw new ResponseHelper.ResponseError(
+							'No encryption key found to decrypt the credentials',
```
Comment (BHesseldieck) [inline packages/cli/src/Server.ts:1999]:
```suggestion
							'Encryption key missing to decrypt credentials',
```

File: packages/cli/src/Server.ts [hunk 11]
```diff
@@ -2189,104 +2193,107 @@ class App {
...
+					const credential = await getCredentialForUser(credentialId, req.user);
 
-				let encryptionKey;
-				encryptionKey = await UserSettings.getEncryptionKey();
-				if (encryptionKey === undefined) {
-					res.status(500).send('No encryption key got found to decrypt the credentials!');
-					return '';
-				}
+					if (!credential) {
+						throw new ResponseHelper.ResponseError('The credential is not known', undefined, 404);
```
Comment (BHesseldieck) [inline packages/cli/src/Server.ts:2211]:
```suggestion
						throw new ResponseHelper.ResponseError('Credential not found', undefined, 404);
```

File: packages/cli/src/Server.ts [hunk 12]
```diff
@@ -2189,104 +2193,107 @@ class App {
...
-					authorizationUri: _.get(oauthCredentials, 'authUrl', '') as string,
-					redirectUri: `${WebhookHelpers.getWebhookBaseUrl()}${
-						this.restEndpoint
-					}/oauth2-credential/callback`,
-					scopes: _.split(_.get(oauthCredentials, 'scope', 'openid,') as string, ','),
-					state: stateEncodedStr,
-				};
+					if (!encryptionKey) {
+						throw new ResponseHelper.ResponseError(
+							'No encryption key found to decrypt the credentials',
```
Comment (BHesseldieck) [inline packages/cli/src/Server.ts:2218]:
```suggestion
							'Encryption key missing to decrypt credentials',
```

**Review by BHesseldieck — CHANGES_REQUESTED** [review]
We should unify error messsages like `'No encryption key got found to decrypt the credentials!'` somehow, for example as constants in a constants file, so we can reuse them on testing 👍🏼

File: packages/cli/src/Server.ts (hunk 8 above)
Comment (ivov) [reply packages/cli/src/Server.ts:1284]: Not my work, but likely because of this:
https://github.com/n8n-io/n8n/pull/2697

**Pull request merged**